### ISSUE TITLE
Allow multiple actions in keybinds

### ISSUE THREAD

**Issue by Consolatis** [labwc/labwc#192]
As written by @ARDiDo in #190:
> Bit of a side note, but we may want to consider adding support for keybindings with multiple actions.

I had a quick look of what would be required to make that happen.
In keyboard.c `handle_keybinding()` function returns after the first match (which is a good thing for a key handler):
https://github.com/labwc/labwc/blob/a47931bba24a8058dfeda4f0c3976bf7f781564d/src/keyboard.c#L68-L73
That means creating a new `struct keybind` based on the current one will not work the same as for mousebinds.
So I'd assume the best way forward would be to convert `struct keybind.action` into a list of ~~strings~~ (see Edit 2) and loop over those in the above handler. This might actually also make sense for mousebinds, it seems cleaner to me.
And if mousebinds are converted as well the loop could instead be done in `action()`.

I never did anything with `wl_list_*` so I'd prefer somebody else to implement that until I get the motivation to look into how the various `wl_*` helpers work.

Edit:
Also that function seems to only check if the modifiers and one of the keysyms match.
So a config like
```
<keybind key="A-q-w-e-r">
  ..
</keybind>
```
Would in fact mean A + any_of(q, w, e, r) pressed. Is this the intended behavior?

Edit 2:
Converting `struct keybind.action` into a list of strings would obviously not work because we require the action arguments like `command` and `menu` as well. So it would have to be a new struct to hold those, something like
```
struct action {
    char *action;
    char *arg;
    struct wl_list link;
}
```

**Comment by johanmalm** [comment]
Multiple actions per {key,mouse}bind is definitely the right thing to implement. `SnapToEdge` is an example of an action that could be simplified and made more generic if we supported multiple actions.

I agree that a `struct action` is needed - not only for a multiple action implementation, but `action(struct view *activator, struct server *server, const char *action, const char *command, uint32_t resize_edges)` is already starting to get a bit big in terms of arguments, and there are lots more variables in the openbox action spec which I suspect we'll want to consider (http://openbox.org/wiki/Help:Actions). For example, see `MoveResizeTo`. This has been on my mind for a while.

I agree the stages ought to be
1. Re-write action() to take a context struct (like you've suggested).
2. Add `struct wl_list actions` to the `keybind` and `mousebind` structs.

Regarding your point on `<keybind key="A-q-w-e-r">`. That should mean all those five keys should be pressed at the same time. 
`Openbox` supports `key chains` (see http://openbox.org/wiki/Help:Bindings) but they are defined using nested xml elements. `labwc` does not (yet) support these. I've never personally used them and am not that bothered.

**Comment by Consolatis** [comment]
If there is a new `struct action` it may make sense to use an Enum instead of `char *action` in above example struct so we don't have to compare strings all the time and can replace that with a single `switch(action->type)`. It also save a bit of CPU and RAM.

**Comment by johanmalm** [comment]
Yep 👍

**Comment by Consolatis** [comment]
Added a first draft in #195

**Comment by johanmalm** [comment]
Closed with 0b45cce6
With thanks :smile:

**Comment by Consolatis** [comment]
I have to thank you (and all the other contributors) for providing a way forward so I can actually switch to wayland as my main desktop in the foreseeable future :smile:
